src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -439,24 +439,29 @@ private async ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFl
                             dataLength = 0;
                         }
 
+                        if (_minResponseDataRate != null)
+                        {
+                            _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
+                        }
+
                         // Don't call TimeFlushUnsynchronizedAsync() since we time this write while also accounting for
                         // flow control induced backpressure below.
                         writeTask = _flusher.FlushAsync();
                     }
                     else if (firstWrite)
                     {
+                        if (_minResponseDataRate != null)
+                        {
+                            _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
+                        }
+
                         // If we're facing flow control induced backpressure on the first write for a given stream's response body,
                         // we make sure to flush the response headers immediately.
                         writeTask = _flusher.FlushAsync();
                     }
 
                     firstWrite = false;
 
-                    if (_minResponseDataRate != null)
-                    {
-                        _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
-                    }
-
                     _unflushedBytes = 0;
```

Comment:
This should only be reset to zero if we've called BytesWrittenToBuffer or _minResponseDataRate is null.